lib/cache/adapters/memory_adapters/queue_adapter.rb

```diff
@@ -4,24 +4,34 @@ module Adapters
       module MemoryAdapters
         # Memory adapter implementation, which stores everything inside queue
         class QueueAdapter
-          def initialize(size)
-            @size = size
+          def initialize(max_size)
+            @max_size = max_size
             @queue = Queue.new
+            @current_size = Concurrent::AtomicFixnum.new(0)
           end
 
           # Adds data to queue in non-blocking mode
           def add_to_queue(data)
+            fail ThreadError if @current_size.value >= @max_size
+
             @queue.push(data)
+
+            @current_size.increment
           end
 
           # Get all items from the queue
           def clear
             items = []
 
-            loop { items << @queue.pop(true) }
+            loop do
+              items << @queue.pop(true)
+
```

Comment:
ok, I think I made a mistake.

I think it CAN happen and will happen that producer will continue adding elements to the queue while this is taking things out right?
If so, I think it's safe to say that we should decrement here every time you take one out.

I'm afraid that 
```ruby
rescue ThreadError
   # Last queue item reached
   # another element is added right between the catch clause and the next time, due to the nature of concurrency
  # and we set by mistake the size to zero, when actually there was one or more elements. 
  @current_size.value = 0

  items
end
```